keps/sig-storage/1472/README.md

```diff
@@ -0,0 +1,1416 @@
+
+### Gathering capacity information
+
+A sidecar, external-provisioner in this proposal, will be extended to
+handle the management of the new objects. This follows the normal
+approach that integration into Kubernetes is managed as part of the
+CSI driver deployment, ideally without having to modify the CSI driver
+itself. This will work without changes for some storage systems while
+others may have to support [a new CSI API
+call](#with-central-controller-generic-solution).
+
+### Pod scheduling
+
+The Kubernetes scheduler watches the capacity information and excludes
+nodes with insufficient remaining capacity when it comes to making
+scheduling decisions for a pod which uses persistent volumes with
+delayed binding. If the driver does not indicate that it supports
+capacity reporting, then the scheduler proceeds just as it does now,
+so nothing changes for existing CSI driver deployments.
+
+## Design Details
+
+### API
+
+#### CSIStorageCapacity
+
+```
+// CSIStorageCapacity stores the result of one CSI GetCapacity call for one
+// driver, one topology segment, and the parameters of one storage class.
+type CSIStorageCapacity struct {
+	metav1.TypeMeta
+	// Standard object's metadata. The name has no particular meaning and just has to
+	// meet the usual requirements (length, characters, unique). To ensure that
+	// there are no conflicts with other CSI drivers on the cluster, the recommendation
+	// is to use csisc-<uuid>.
+	//
+	// Objects are not namespaced.
+	//
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+
+    // Spec contains the properties that describe one capacity value.
+    Spec CSIStorageCapacitySpec
+}
+
+// CSIStorageCapacitySpec contains the properties that describe one capacity value.
+type CSIStorageCapacitySpec struct {
+	// The CSI driver that provides the storage.
+	// This must be the string returned by the CSI GetPluginName() call.
+	DriverName string
+
+    // NodeTopology defines which nodes have access to the storage for which
+    // capacity was reported. If not set, the storage is accessible from all
+    // nodes in the cluster.
+	// +optional
+	NodeTopology *v1.NodeSelector
+
+	// The storage class name of the StorageClass which provided the
+    // additional parameters for the GetCapacity call.
+	StorageClassName string
+
+	// Capacity is the value reported by the CSI driver in its GetCapacityResponse.
+    // Depending on how the driver is implemented, this might be the total
+    // size of the available storage which is only available when allocating
+    // multiple smaller volumes ("total available capacity") or the
+    // actual size that a volume may have ("maximum volume size").
+    //
+    // It is optional because in the future, other ways of describing capacity
+    // might get added. If not set, the scheduler will simply ignore the
+    // CSIStorageCapacity object.
+    //
+	// +optional
```

Comment:
This is meant to enable future extensions. Once we introduce "total capacity" and "maximum volume size", this API would become:
```
Capacity *resource.Quantity
TotalCapacity *resource.Quantity
MaximumVolumeSize *resource.Quantity
```

At least one will be required then.

Let me change the "optional" into "required", because right now it is.